src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -2515,6 +2517,40 @@ create_exp_macro(Exp, exp)
 create_exp_macro(Exp2, exp2)
 create_exp_macro(Expm1, expm1)
 
+namespace Input {
+
+    static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
+            diag::Diagnostics& diagnostics) {
+        ASRUtils::require_impl(x.n_args <= 1,
+            "ASR Verify: Call `input` takes 0 or 1 argument",
+            x.base.base.loc, diagnostics);
+        if (x.n_args == 1) {
+            ASR::ttype_t *type = ASRUtils::expr_type(x.m_args[0]);
+            ASRUtils::require_impl(ASRUtils::is_character(*type),
+                "ASR Verify: Argument to `input` must be of string type",
+                x.base.base.loc, diagnostics);
+        }
+    }
+
+    static inline ASR::asr_t* create_Input(Allocator& al, const Location& loc,
+            Vec<ASR::expr_t*>& args,
+            const std::function<void (const std::string &, const Location &)> err) {
+        if (args.n != 1) {
+            err("input() takes 0 or 1 argument", loc);
+        }
+        if (!ASRUtils::is_character(*ASRUtils::expr_type(args[0]))) {
+            err("Argument `prompt` to input() must be of type 'str'",
+                loc);
+        }
+        std::string dummy_string = "";
+        ASR::ttype_t* return_type = expr_type(StringConstant(dummy_string, character(dummy_string.length())));
+        return ASR::make_IntrinsicScalarFunction_t(al, loc,
+            static_cast<int64_t>(IntrinsicScalarFunctions::Input),
+            args.p, args.n, 0, return_type, nullptr);
+    }
```

Comment:
You should also define an `instantiate_Input()` function here which will carry the definition of the input function.
In the input function definition, that is in its function body, you can have a statement `visit_FileRead()` (with `m_unit` as `nullptr`) that reads into the `result` variable which is the return value for your `input()` function. This return value will contain the value read from `stdin`.

Reply:
@Shaikh-Ubaid I am facing some problems calling the `visit_FileRead()`. It appears to have been declared in the `ast_to_*` files and as it turns out, is not being used anywhere in the whole code-base. I thought maybe I could call it using the `ASRToLLVMVisitor`, but again noticed that it was not being used anywhere too. An include error pops up when including the file which contains the class, stating dependency issues with the file.

But most important of all, `ASRToLLVMVisitor::visit_FileRead()` requires a `ASR::FileRead_t` object. It has got many required arguments. What should I pass to them? 

Here is my code:
```c++
static inline ASR::expr_t* instantiate_Input(Allocator &al, const Location &loc,
        SymbolTable /*scope*/, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
        std::string func_name = "_lcompilers_input_";
        declare_basic_variables(func_name);
        fill_func_arg("prompt", arg_types[0]);
        auto result = declare(fn_name, return_type, ReturnVar);
        ASRToLLVMVisitor::visit_FileRead(ASR::make_FileRead_t(al, loc, 0, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, 0));
    }
```

Reply:
>What should I pass to them?

Just pass the unneeded parameters as `nullptr`.

>ASRToLLVMVisitor::visit_FileRead(ASR::make_FileRead_t(al, loc, 0, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, 0));

You need not call the `ASRToLLVMVisitor::visit_FileRead()` here. You just need to construct ASR which contains the FileRead node. This ASR node will automatically be visited in `visit_FileRead()` during ASR->LLVM construction.

For example, you updated instantiate_Input() would like 
```console
auto result = declare(fn_name, return_type, ReturnVar);
Vec<ASR::expr_t*> read_values;
read_values.reserve(al, 0);
read_values.push_back(al, result);
body.push_back(ASRUtils::STMT(ASR::make_FileRead_t(al, loc, 0, nullptr, nullptr, nullptr, nullptr, nullptr, read_values.p, read_values.n)));
```

Reply:
@Shaikh-Ubaid `read_values.reserve(al, 0);`?

Reply:
>@Shaikh-Ubaid read_values.reserve(al, 0);?

It is not the perfect code but just a guideline for you. Please update it as needed.

Checkout implementations for other `instantiate_*()` to understand how to define an intrinsic function in ASR intrinsic registry. In LFortran we have many intrinsics implemented, so you can find many examples there.

Reply:
@Shaikh-Ubaid Thank you very much for all your guidance! I have learnt a lot!